solidity/contracts/bridge/Frauds.sol

```diff
@@ -0,0 +1,577 @@
+        /* solhint-enable avoid-low-level-calls */
+
+        bytes memory compressedWalletPublicKey = EcdsaLib.compressPublicKey(
+            walletPublicKey.slice32(0),
+            walletPublicKey.slice32(32)
+        );
+        bytes20 walletPubKeyHash = bytes20(compressedWalletPublicKey.hash160());
+
+        emit FraudChallengeDefeated(
+            walletPubKeyHash,
+            sighash,
+            signature.v,
+            signature.r,
+            signature.s
+        );
+    }
+
+    /// @notice Notifies about defeat timeout for the given fraud challenge.
+    ///         Can be called only if there was a fraud challenge identified by
+    ///         the provided `walletPublicKey`, `sighash` and signature
+    ///         (represented by `r`, `s` and `v`) and it was not defeated on time.
+    ///         The amount of time that needs to pass after a fraud challenge is
+    ///         reported is indicated by the `challengeDefeatTimeout`. After a
+    ///         successful fraud challenge defeat timeout notification the fraud
+    ///         challenge is closed, the stake of each operator is slashed, the
+    ///         ether deposited is returned to the challenger and the challenger
+    ///         is rewarded.
+    /// @param walletPublicKey The public key of the wallet in the uncompressed
+    ///        and unprefixed format (64 bytes)
+    /// @param sighash The hash that was used to produce the ECDSA signature
+    ///        that is the subject of the fraud claim. This hash is constructed
+    ///        by applying double SHA-256 over a serialized subset of the
+    ///        transaction. The exact subset used as hash preimage depends on
+    ///        the transaction input the signature is produced for. See BIP-143
+    ///        for reference
+    /// @param signature Bitcoin signature in the R/S/V format
+    function notifyFraudChallengeDefeatTimeout(
+        Data storage self,
+        bytes memory walletPublicKey,
+        bytes32 sighash,
+        Frauds.RSVSignature calldata signature
+    ) external {
+        uint256 challengeKey = uint256(
+            keccak256(
+                abi.encodePacked(
+                    walletPublicKey,
+                    sighash,
+                    signature.v,
+                    signature.r,
+                    signature.s
+                )
+            )
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+        require(challenge.reportedAt > 0, "Fraud challenge does not exist");
+        require(!challenge.closed, "Fraud challenge is closed");
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            block.timestamp >=
+                challenge.reportedAt + self.challengeDefeatTimeout,
+            "Fraud challenge defeat period did not time out yet"
+        );
+
+        // TODO: Slash the wallet
+        // TODO: Reward the challenger
```

Comment:
And `// TODO: Terminate the wallet` :)

Reply:
Or maybe just `// TODO: Call `notifyFraud` from `Wallets` library`

Reply:
Done in b0eb1f3c825cee422c76c8b0d5fdccdf88397f7a